### PR TITLE
Alert on KubePersistentVolumeRunningOutOfInodes

### LINKED ISSUE #426: Alerts on inodes running out
Hi,
It would be nice to have alerts on when PVCs are about to run out of free inodes. 

For example by using these metrics:
- `kubelet_volume_stats_inodes_free`
- `kubelet_volume_stats_inodes`

There are already other storage alerts, but they only look at bytes usage (or errors):
https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/8ede288c0e3f72513931e46445329ab084f79daa/alerts/storage_alerts.libsonnet

There's a dashboard that shows the inodes usage, but wouldn't it be nice to get alerted about it? 

Skärmavbild 2020-05-20 kl  10 40 44

### PR DESCRIPTION
**Pull request by MPV**
Based on `KubePersistentVolumeFillingUp`, but using these metrics instead:
- kubelet_volume_stats_inodes_free
- kubelet_volume_stats_inodes

Fixes #426.

### CODE REVIEW

**Comment by MPV** [comment]
@povilasv (as you looked into #426)

File: alerts/storage_alerts.libsonnet [hunk 1]
```diff
@@ -51,6 +51,41 @@
               message: 'Based on recent sampling, the PersistentVolume claimed by {{ $labels.persistentvolumeclaim }} in Namespace {{ $labels.namespace }} is expected to fill up within four days. Currently {{ $value | humanizePercentage }} is available.',
             },
           },
+          {
+            alert: 'KubePersistentVolumeRunningOutOfInodes',
```
Comment (povilasv) [inline alerts/storage_alerts.libsonnet:55]:
```suggestion
            alert: 'KubePersistentVolumeLowOnInodes',
```

File: alerts/storage_alerts.libsonnet [hunk 2]
```diff
@@ -51,6 +51,41 @@
...
+            'for': '1m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              message: 'The PersistentVolume claimed by {{ $labels.persistentvolumeclaim }} in Namespace {{ $labels.namespace }} only has {{ $value | humanizePercentage }} free inodes.',
+            },
+          },
+          {
+            alert: 'KubePersistentVolumeRunningOutOfInodes',
```
Comment (povilasv) [inline alerts/storage_alerts.libsonnet:71]:
```suggestion
            alert: 'KubePersistentVolumeLowOnInodes',
```

File: alerts/storage_alerts.libsonnet [hunk 3]
```diff
@@ -51,6 +51,41 @@
...
+            expr: |||
+              kubelet_volume_stats_inodes_free{%(prefixedNamespaceSelector)s%(kubeletSelector)s}
+                /
+              kubelet_volume_stats_inodes{%(prefixedNamespaceSelector)s%(kubeletSelector)s}
+                < 0.03
+            ||| % $._config,
+            'for': '1m',
+            labels: {
+              severity: 'critical',
```
Comment (metalmatze) [inline alerts/storage_alerts.libsonnet:64]: I'm not sure we should have the critical one at all.
`severity: warning` is totally fine :+1: I'm a bit concerned, cause there might be some application that's running totally fine with less than `0.03` inodes left.
This being a cause and not even a symptom and then paging to wake up people is probably too much. If it really becomes a problem, the alerting on your actual application should page and wake one up.

Comment (paulfantom) [reply alerts/storage_alerts.libsonnet:64]: We do have the same alerting structure for [running out of space](https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/master/alerts/storage_alerts.libsonnet#L19-L34) and this is also [accepted pattern in node_exporter mixin](https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/master/alerts/storage_alerts.libsonnet#L19-L34). I think we should be consistent and either reduce severity here and in `KubePersistentVolumeFillingUp` or accept that running out of space/inodes is one of the exceptions where alerting on cause is something that is excused due to lack of good symptom.

Comment (metalmatze) [reply alerts/storage_alerts.libsonnet:64]: Ok, I guess I can live with the fact to have this for the sake of consistency. It doesn't mean I'm entirely keen on the idea. :upside_down_face: In this case, feel free to go ahead and ignore my comment. :grin:

**Comment by arajkumar** [comment]
Looks like the PR is waiting for @MPV to address few review concerns. @stefanandres would you like to take it forward?

**Comment by stefanandres** [comment]
This issue can be closed, we've put https://github.com/kubernetes-monitoring/kubernetes-mixin/pull/761 forward which has been merged.

**Comment by MPV** [comment]
Wonderful, thanks! 🙏🏻

**Pull request closed without merging**